### ISSUE TITLE
Improve performance & lower memory usage

### ISSUE THREAD

**Issue by DiegoRBaquero** [webtorrent/bittorrent-tracker#175]
Although we have made several improvements and clear memory leaks, everything can always be improved.

In order for the webtorrent-webrtc-based projects to reach more users easily, the trackers need to be able to scale with greater performance.

At the time of writing, 1K connected users go for about 500MB of RAM and each connected user is around 0.45MB (BTorrent's tracker)

**Comment by feross** [comment]
@DiegoRBaquero Agreed, this is important.

From your testing, do you know which objects are using the most memory?

**Comment by DiegoRBaquero** [comment]
@feross Took a heapdump:  https://webseed.btorrent.xyz/heapdump-1587638310.868854.heapsnapshot

Sintel in the tracker memory:
![sintel](https://www.dropbox.com/s/uy05bjtn849x1aq/Screenshot%202016-10-14%2000.48.43.jpg?dl=1)

Other ways to approach this problem: 

- REST tracker over HTTP2
- FIFO peers stack

**Comment by DiegoRBaquero** [comment]
@feross

Tested uWS and it was a HUGE improvement. The only thing lost is the real IP address (which is never used, actually protects peers' privacy) and the full headers of the request (Again, protects peers' privacy from trackers' operators but would not allow filtering or blocking).

@alxhotel @jakefb You might want to check this out.

**Comment by yciabaud** [comment]
That looks interesting! Nice work! Have you looked at the heap when using µWS?
Do you thing we could make significant memory optimizations in our codebase?

**Comment by DiegoRBaquero** [comment]
@yciabaud I didn't even bother, but I could get a heapshot if you'd like to see if there's anything else we can improve.

We are only left with clearing 1-dead-peer-only zombie infoHashes (https://github.com/feross/bittorrent-tracker/issues/164)

By the way, guys, could I get comments on: https://medium.com/@diegorbaquero/%C2%B5ws-as-your-next-websocket-library-d34209686357#.twgf337zc

Thanks

**Comment by lpinca** [comment]
@DiegoRBaquero interesting. Did you try to disable permessage-deflate in both `ws` and `uws`? I know that the overhead for `ws` is pretty big. Also to reduce memory usage a little you can try null the `upgradeReq` object when the 'connection' event is emitted, `uws` does this under the hood.

```js
wss.on('connection', (ws) => {
  ws.upgradeReq = null;
});
```

I'm actually not sure if this works as expected.
It might also be interesting to see if installing `ws` from master makes any difference. There are some improvements including breaking changes that have not been released yet.

```
npm i websockets/ws bufferutil
```

**Comment by lpinca** [comment]
I also see that the `clientTracking` option for `ws` is set to `true` (default value). It seems that this lib uses its own way to track clients so you may want to set it to `false`.

**Comment by DiegoRBaquero** [comment]
@lpinca Had not try either the options you describe, had I seen them in the docs por better performance I would have tried them before.

I'll be monitoring uWS performance for the weekend and PR the change on monday if everything is cool.

**Comment by alexhultman** [comment]
> That looks interesting! Nice work! Have you looked at the heap when using µWS?
Do you thing we could make significant memory optimizations in our codebase?

µWS doesn't use the JS memory at all (except for the small JS wrapper that is) so that heapdumper won't show real memory usage.

> I also see that the clientTracking option for ws is set to true (default value). It seems that this lib uses its own way to track clients so you may want to set it to false.

µWS has built in "clientTracking" that cannot be turned off.

> Did you try to disable permessage-deflate in both ws and uws? I know that the overhead for ws is pretty big. Also to reduce memory usage a little you can try to null the upgradeReq object when the 'connection' event is emitted, uws does this under the hood.

µWS has permessage-deflate enabled by default, just like ws. Removing upgradeReq from ws does *not* improve memory usage, it actually slightly increases it (whatever V8 is doing, it is not doing the correct thing). The reason why µWS stresses so hard on removing upgradeReq is because it holds the net.Socket object which is *massive*. µWS does not use the net.Socket at all, so removing upgradeReq means a 600 mb memory increase for 1 million connections. Again: this does not apply to ws.

TLDR; https://medium.com/@diegorbaquero/%C2%B5ws-as-your-next-websocket-library-d34209686357#.twgf337zc is 100% legit. Nice post dude, but I would recommend disabling permessage-deflate for a lot better throughput. I do agree permessage-deflate should not be enabled by default.

**Comment by alexhultman** [comment]
> I'm actually not sure if this works as expected.

Well, you certainly *seem to know* a lot about the upgradeReq since you have been insinuating all over that `ws` could *easily* scale to µWS levels if upgradeReq was removed. Obviously you have *never actually tried* this, which you also show here.

Bravo :clap: :clap:

**Comment by alexhultman** [comment]
@lpinca Maybe you should listen to me when I tell you what is the best usage of µWS? Instead of acting like some sort of dictator inside of the Primus camp? I have told you since version 0.4.0 that *you have to remove the upgradeReq in the µWS transformer* - not only because it massively hurts memory usage to leave it in, but also the fact that *it holds references to deleted memory that should not be used by the app developer*.

Did you listen? Doesn't seem like it.

**Comment by lpinca** [comment]
@alexhultman

> I'm actually not sure if this works as expected.

I'm not sure because there might be other references that prevent the `upgradeReq` from being GC'd.

> Well, you certainly seem to know a lot about the upgradeReq since you have been insinuating all over that ws could easily scale to µWS levels if upgradeReq was removed.

I've never said that `ws` could scale at the same levels of uws, if I did show me where please. I only said that your numbers are wrong and they still are. permessage-deflate seem to have serious issues in `ws`, see https://github.com/websockets/ws/issues/804.

>  Instead of acting like some sort of dictator inside of the Primus camp? I have told you ages ago that you have to remove the upgradeReq in the µWS transformer

This make me sad... Did you actually read why we can't do that? I guess not.
Want to discuss about Primus? I'm all for it but it's not relevant to this issue. Let's do it somewhere else.

I've only suggested how to improve ws performance here.

**Comment by yciabaud** [comment]
@DiegoRBaquero can we move forward on this? Do you still think we should switch to uWS? Did you test disabling permessage-deflate?

**Comment by DiegoRBaquero** [comment]
@alxhotel @jakefb Could you guys update to latest and test memory consumption with per message deflate and client tracking off in ws?

**Comment by alxhotel** [comment]
@DiegoRBaquero Done! Results from the last 12 hours are aprox: 5x less memory consumption and 3x more network bandwidth.

**Comment by DiegoRBaquero** [comment]
@alxhotel Were there any restarts due to errors? 

I think I caused a restart when loading /stats, which evicts old peers.

**Comment by alexhultman** [comment]
If you want a sweetspot between low memory usage and low network usage you can spend some time optimizing the WebTorrent protocol. I assume it is based on JSON right now and that is very ineffective vs something binary. I bet you can improve it a ton in size.

**Comment by feross** [comment]
@alexhultman We could try switching back to raw `bencoding` since that's what the other tracker implementations (http and udp) use. But we would need to consider a transition period where we supported both bencoding/json for several months until clients are updated.

**Comment by alxhotel** [comment]
@DiegoRBaquero Yep. See issue #205

**Comment by Simplici** [comment]
I seeded a 4G file, it had token almost 8G memory.
